stan/math/gpu/matrix_gpu.hpp

```diff
@@ -0,0 +1,283 @@
+    oclBuffer_ = a.buffer();
+    return *this;
+  }
+};
+
+/**
+ * Copies the source Eigen matrix to
+ * the destination matrix that is stored
+ * on the GPU.
+ *
+ * @tparam T type of data in the Eigen matrix
+ * @param src source Eigen matrix
+ * @param dst destination matrix on the GPU
+ *
+ * @throw <code>std::invalid_argument</code> if the
+ * matrices do not have matching dimensions
+ *
+ */
+template <typename T, int R, int C>
+void copy(const Eigen::Matrix<T, R, C>& src, matrix_gpu& dst) {
+  check_size_match("copy (Eigen -> GPU)", "src.rows()", src.rows(),
+                   "dst.rows()", dst.rows());
+  check_size_match("copy (Eigen -> GPU)", "src.cols()", src.cols(),
+                   "dst.cols()", dst.cols());
+  if (src.size() > 0) {
+    cl::CommandQueue queue = opencl_context.queue();
+    try {
+      /** writes the contents of src to the OpenCL buffer
+       * starting at the offset 0
+       * CL_TRUE denotes that the call is blocking
+       * We do not want to execute any further kernels
+       * on the device until we are sure that the data is transferred)
+       */
+      queue.enqueueWriteBuffer(dst.buffer(), CL_TRUE, 0, sizeof(T) * dst.size(),
+                               src.data());
+    } catch (const cl::Error& e) {
+      check_opencl_error("copy Eigen->GPU", e);
+    }
+  }
+}
+
+/**
+ * Copies the source matrix that is stored
+ * on the GPU to the destination Eigen
+ * matrix.
+ *
+ * @tparam T type of data in the Eigen matrix
+ * @param src source matrix on the GPU
+ * @param dst destination Eigen matrix
+ *
+ * @throw <code>std::invalid_argument</code> if the
+ * matrices do not have matching dimensions
+ *
+ */
+template <typename T, int R, int C>
+void copy(matrix_gpu& src, Eigen::Matrix<T, R, C>& dst) {
+  check_size_match("copy (GPU -> Eigen)", "src.rows()", src.rows(),
+                   "dst.rows()", dst.rows());
+  check_size_match("copy (GPU -> Eigen)", "src.cols()", src.cols(),
+                   "dst.cols()", dst.cols());
+  if (src.size() > 0) {
+    cl::CommandQueue queue = opencl_context.queue();
+    try {
+      /** reads the contents of the OpenCL buffer
+       * starting at the offset 0 to the Eigen
+       * matrix
+       * CL_TRUE denotes that the call is blocking
+       * We do not want to execute any further kernels
+       * on the device until we are sure that the data is transferred)
+       */
+      queue.enqueueReadBuffer(src.buffer(), CL_TRUE, 0, sizeof(T) * dst.size(),
+                              dst.data());
+    } catch (const cl::Error& e) {
+      check_opencl_error("copy GPU->Eigen", e);
+    }
+  }
+}
+
+/**
+ * Copies the source matrix to the
+ * destination matrix. Both matrices
+ * are stored on the GPU.
+ *
+ * @param src source matrix
+ * @param dst destination matrix
+ *
+ * @throw <code>std::invalid_argument</code> if the
+ * matrices do not have matching dimensions
+ *
+ */
+inline void copy(matrix_gpu& src, matrix_gpu& dst) {  // NOLINT
```

Comment:
why does this need a `// NOLINT` comment? If it's absolutely necessary, please include the specific exception to cpplint.

Reply:
Removing the //NOLINT and running `make cpplint`

```
stan/math/gpu/matrix_gpu.hpp(252):  Add #include <algorithm> for copy  [build/include_what_you_use] [4]
```

Is there some way to turn the search for that word? I'm guessing we want that check elsewhere tho' so I can include a comment here on why we have a no lint